compiler/rustc_middle/src/mir/mono.rs

```diff
@@ -103,17 +102,13 @@ impl<'tcx> MonoItem<'tcx> {
                 // inlined function. If we're inlining into all CGUs then we'll
                 // be creating a local copy per CGU.
                 if generate_cgu_internal_copies {
-                    return InstantiationMode::LocalCopy;
-                }
-
-                // Finally, if this is `#[inline(always)]` we're sure to respect
-                // that with an inline copy per CGU, but otherwise we'll be
-                // creating one copy of this `#[inline]` function which may
-                // conflict with upstream crates as it could be an exported
-                // symbol.
-                match tcx.codegen_fn_attrs(instance.def_id()).inline {
```

Comment:
Am I correct to assume that the warning this comment makes regarding conflicts with upstream crates isn't an issue because we check for explicit linkage above?

Reply:
The concern as I understand it is that you have a crate tree that looks like:

```
      A
      | \
      |  B
      | / 
      C  
```

Where `A` has some function `fn foo<T>() { .. }` which `B` and `C` both instantiate for some type (ie, `B` and `C` both use `A::foo<u8>()`). Because we want to share the same instantiation of `A::foo<u8>()` within the crate `B`'s codegen units, it will not be a `LocalCopy` or marked private, then the symbol will be exported as part of `B` (although with visibility `Hidden`). Of course, the same logic applies to `C` and so now there will be two symbols in different crates with the same. 

This is where `may_conflict: true` comes in. When this is set, the mangled name for this instantiation includes information about the instantiating crate which means the symbols are no longer the same and thus linking will succeed. 

See also 

https://github.com/rust-lang/rust/blob/3c45bff23dfbb13d9d28afe5dba57c028ba156da/compiler/rustc_symbol_mangling/src/lib.rs#L220-L233